model/userdata.go

```diff
@@ -0,0 +1,147 @@
+package model
+
+import (
+	"github.com/coreos/coreos-cloudinit/config/validate"
+	"github.com/kubernetes-incubator/kube-aws/filereader/texttemplate"
+	"github.com/kubernetes-incubator/kube-aws/gzipcompressor"
+
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"text/template"
+)
+
+// UserDataPartType defines type of the content in a given part
+// it mostly affects how part should be validated
+type UserDataPartType int
+
+const (
+	USERDATA_SHELL UserDataPartType = iota
+	USERDATA_COREOS_CLOUDINIT
+	USERDATA_YAML
+)
+
+// UserData represents userdata which might be split across multiple storage types
+type UserData struct {
+	Parts map[string]UserDataPart
+}
+
+type UserDataPart struct {
+	Type     UserDataPartType
+	Asset    Asset
+	tmpl     *template.Template
+	tmplData interface{}
+}
```

Comment:
I'd try introducing a new type `userDataPart` which has everything currently implemented for `UserDataPart` but `.validate()`. Then, it could be embedded into `UserDataInstancePart` and `UserDataS3Part` like:

```golang
type userDataPart struct {
	Asset    Asset
	tmpl     *template.Template
	tmplData interface{}
}

type UserDataInstancePart struct {
	userDataPart
}

type UserDataS3Part struct {
	userDataPart
}

func (self userDataPart) Base64(compress bool, extra ...map[string]interface{}) (string, error) {
	//...
}

func (self userDataPart) Template(extra ...map[string]interface{}) (string, error) {
	//...
}

func (self UserDataInstancePart) validate(content []byte) error {
	//...
}

func (self UserDataS3Part) validate(content []byte) error {
	//...
}
```

I believe this is still arguably a trade-off between verbosity and testability(?). WDYT?